streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java

```diff
@@ -1281,6 +1281,9 @@ private boolean nodeGroupContainsGlobalSourceNode(final Set<String> allNodesOfGr
         @Override
         public int compare(final TopologyDescription.Node node1,
                            final TopologyDescription.Node node2) {
+            if (node1.equals(node2)) {
```

Comment:
While debugging the issue, I figure that the use comparators don't do deep comparison. This is the corresponding fix. (Similar below)